File: Examples/Advanced/Export-Device-Interfaces/nodeJS/export-device-interfaces.mjs
#!/usr/bin/env node

import { randomUUID } from "node:crypto";
import { mkdir, open, readFile, rename, rm } from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import { fileURLToPath, pathToFileURL } from "node:url";

export const INTERFACE_FIELDS = [
  "network", "deviceIpAddress", "hostname", "interfaceId", "interfaceIndex",
  "name", "ifName", "type", "description", "comment", "macAddress", "mtu",
  "speed", "adminUp", "vrfName", "ipAddresses",
];
export const FAILURE_FIELDS = ["network", "deviceIpAddress", "hostname", "error"];

export class ExampleError extends Error {}

export function parseNetworks(value) {
  const networks = value.split(",").map((item) => item.trim()).filter(Boolean);
  if (!networks.length) throw new ExampleError("NETLD_NETWORKS must contain at least one managed network.");
  return networks;
}

function csvCell(value) {
  if (value == null) return "";
  const text = String(value);
  return /[",\r\n]/.test(text) ? `"${text.replaceAll('"', '""')}"` : text;
}

function csvRow(values) {
  return `${values.map(csvCell).join(",")}\n`;
}

export function flattenIpAddresses(interfaceData) {
  return (interfaceData.ipAddresses || [])
    .filter((address) => String(address.ipAddress || "").trim())
    .map((address) => address.cidrPrefix == null ? String(address.ipAddress) : `${address.ipAddress}/${address.cidrPrefix}`)
    .join(";");
}

export function interfaceRow(device, interfaceData) {
  return {
    network: device.network, deviceIpAddress: device.ipAddress, hostname: device.hostname,
    interfaceId: interfaceData.id, interfaceIndex: interfaceData.index,
    name: interfaceData.name, ifName: interfaceData.ifName, type: interfaceData.type,
    description: interfaceData.description, comment: interfaceData.comment,
    macAddress: interfaceData.macAddress, mtu: interfaceData.mtu, speed: interfaceData.speed,
    adminUp: interfaceData.adminUp, vrfName: interfaceData.vrfName,
    ipAddresses: flattenIpAddresses(interfaceData),
  };
}

async function loadDotEnv(file) {
  let text;
  try {
    text = await readFile(file, "utf8");
  } catch (error) {
    if (error.code === "ENOENT") return;
    throw error;
  }
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith("#") || !line.includes("=")) continue;
    const offset = line.indexOf("=");
    process.env[line.slice(0, offset).trim()] = line.slice(offset + 1).trim().replace(/^(['"])(.*)\1$/, "$2");
  }
}

async function loadConfig(directory, envFile) {
  await loadDotEnv(envFile);
  for (const name of ["NETLD_BASE_URL", "NETLD_API_KEY"]) {
    if (!process.env[name]?.trim()) throw new ExampleError(`Set ${name} in the environment file.`);
  }
  const pageSize = Number.parseInt(process.env.NETLD_PAGE_SIZE || "500", 10);
  if (!Number.isInteger(pageSize) || pageSize <= 0) throw new ExampleError("NETLD_PAGE_SIZE must be a positive integer.");
  return {
    baseUrl: process.env.NETLD_BASE_URL.replace(/\/$/, ""), apiKey: process.env.NETLD_API_KEY,
    networks: parseNetworks(process.env.NETLD_NETWORKS || "Default"), pageSize,
    outputFile: path.resolve(directory, process.env.NETLD_OUTPUT_FILE || "interfaces.csv"),
    failureFile: path.resolve(directory, process.env.NETLD_FAILURE_FILE || "interface-failures.csv"),
    scheme: process.env.NETLD_SEARCH_SCHEME?.trim() || "ipAddress",
    query: process.env.NETLD_SEARCH_QUERY || "",
  };
}

class NetLDClient {
  constructor(baseUrl, apiKey, timeout = 30000) {
    this.baseUrl = baseUrl;
    this.apiKey = apiKey;
    this.timeout = timeout;
    this.cookies = "";
  }

  headers(extra = {}) {
    return { Authorization: `Bearer ${this.apiKey}`, ...(this.cookies ? { Cookie: this.cookies } : {}), ...extra };
  }

  async request(options = {}) {
    let response;
    try {
      response = await fetch(`${this.baseUrl}/rest`, {
        redirect: "manual", signal: AbortSignal.timeout(this.timeout), ...options,
        headers: this.headers(options.headers),
      });
    } catch {
      throw new ExampleError(`Could not reach ${this.baseUrl}.`);
    }
    if (response.status >= 300 && response.status < 400) {
      throw new ExampleError(`Request redirected to ${response.headers.get("location") || ""}.`);
    }
    if (!response.ok) throw new ExampleError(`Request failed with HTTP ${response.status}.`);
    const setCookies = response.headers.getSetCookie?.() || splitSetCookie(response.headers.get("set-cookie") || "");
    if (setCookies.length) this.cookies = setCookies.map((item) => item.split(";", 1)[0]).join("; ");
    return response;
  }

  async login() { await this.request(); }

  async call(method, parameters = {}) {
    const response = await this.request({
      method: "POST", headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ jsonrpc: "2.0", method, params: parameters, id: randomUUID() }),
    });
    const data = await response.json();
    if (data.error) throw new ExampleError(`${method} failed: ${JSON.stringify(data.error)}`);
    return data.result;
  }

  async searchInventory(networks, scheme, query, offset, pageSize) {
    const result = await this.call("Inventory.search", {
      network: networks, scheme, query: query.endsWith("\n") ? query : `${query}\n`,
      pageData: { offset, pageSize }, sortColumn: "ipAddress", descending: false,
    });
    if (!result || typeof result !== "object") throw new ExampleError("Inventory.search returned no page data.");
    return result;
  }

  async getDeviceInterfaces(network, ipAddress) {
    const result = await this.call("Inventory.getDeviceInterfaces", { network, ipAddress });
    if (result == null) return [];
    if (!Array.isArray(result)) throw new ExampleError("Inventory.getDeviceInterfaces returned an invalid collection.");
    return result;
  }
}

function splitSetCookie(header) {
  return header ? header.split(/,(?=\s*[^;=]+=[^;]+)/) : [];
}

export async function exportInterfaces(client, config) {
  await mkdir(path.dirname(config.outputFile), { recursive: true });
  await mkdir(path.dirname(config.failureFile), { recursive: true });
  const interfaceTemp = path.join(path.dirname(config.outputFile), `.${path.basename(config.outputFile)}.${randomUUID()}.tmp`);
  const failureTemp = path.join(path.dirname(config.failureFile), `.${path.basename(config.failureFile)}.${randomUUID()}.tmp`);
  let interfaceHandle;
  let failureHandle;
  let deviceCount = 0;
  let interfaceCount = 0;
  let failureCount = 0;
  try {
    interfaceHandle = await open(interfaceTemp, "wx");
    failureHandle = await open(failureTemp, "wx");
    await interfaceHandle.write(csvRow(INTERFACE_FIELDS));
    await failureHandle.write(csvRow(FAILURE_FIELDS));
    let offset = 0;
    let total;
    while (true) {
      const page = await client.searchInventory(config.networks, config.scheme, config.query, offset, config.pageSize);
      const devices = page.devices || [];
      if (!Array.isArray(devices)) throw new ExampleError("Inventory.search returned an invalid devices collection.");
      for (const device of devices) {
        deviceCount += 1;
        let interfaces;
        try {
          interfaces = await client.getDeviceInterfaces(device.network, device.ipAddress);
        } catch (error) {
          const failure = { network: device.network, deviceIpAddress: device.ipAddress, hostname: device.hostname, error: error.message };
          await failureHandle.write(csvRow(FAILURE_FIELDS.map((field) => failure[field])));
          failureCount += 1;
          continue;
        }
        for (const interfaceData of interfaces) {
          const row = interfaceRow(device, interfaceData);
          await interfaceHandle.write(csvRow(INTERFACE_FIELDS.map((field) => row[field])));
          interfaceCount += 1;
        }
      }
      const returnedPageSize = Number(page.pageSize || config.pageSize);
      if (!Number.isInteger(returnedPageSize) || returnedPageSize <= 0) throw new ExampleError("Inventory.search returned an invalid page size.");
      if (total == null && page.total != null) total = Number(page.total);
      if (total != null && offset + devices.length >= total) break;
      if (total == null && devices.length < returnedPageSize) break;
      if (!devices.length) throw new ExampleError("Inventory.search returned an empty page before the reported total.");
      offset += returnedPageSize;
    }
    await interfaceHandle.close(); interfaceHandle = undefined;
    await failureHandle.close(); failureHandle = undefined;
    await rename(interfaceTemp, config.outputFile);
    await rename(failureTemp, config.failureFile);
    return { deviceCount, interfaceCount, failureCount };
  } catch (error) {
    await interfaceHandle?.close().catch(() => {});
    await failureHandle?.close().catch(() => {});
    await rm(interfaceTemp, { force: true }).catch(() => {});
    await rm(failureTemp, { force: true }).catch(() => {});
    throw error;
  }
}

export async function main() {
  const directory = path.dirname(fileURLToPath(import.meta.url));
  const envOffset = process.argv.indexOf("--env");
  const envFile = envOffset >= 0 ? path.resolve(process.argv[envOffset + 1]) : path.join(directory, ".env");
  if (envOffset >= 0 && !process.argv[envOffset + 1]) throw new ExampleError("--env requires a file path.");
  const config = await loadConfig(directory, envFile);
  const client = new NetLDClient(config.baseUrl, config.apiKey);
  await client.login();
  const result = await exportInterfaces(client, config);
  console.log(`Processed ${result.deviceCount} devices and wrote ${result.interfaceCount} interfaces to ${config.outputFile}`);
  console.log(`Wrote ${result.failureCount} device lookup failures to ${config.failureFile}`);
  return result.failureCount ? 2 : 0;
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main().then((code) => { process.exitCode = code; }).catch((error) => {
    console.error(`Error: ${error.message}`);
    process.exitCode = 1;
  });
}

File: Examples/Advanced/Export-Device-Interfaces/Python/export_device_interfaces.py
#!/usr/bin/env python3
"""Export collected device interfaces from netLD/ThirdEye to CSV."""

from __future__ import annotations

import argparse
import csv
import json
import os
import sys
import tempfile
import uuid
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterator

import requests
from dotenv import load_dotenv


INTERFACE_FIELDS = (
    "network", "deviceIpAddress", "hostname", "interfaceId", "interfaceIndex",
    "name", "ifName", "type", "description", "comment", "macAddress", "mtu",
    "speed", "adminUp", "vrfName", "ipAddresses",
)
FAILURE_FIELDS = ("network", "deviceIpAddress", "hostname", "error")


class ExampleError(RuntimeError):
    pass


def parse_networks(value: str) -> list[str]:
    networks = [item.strip() for item in value.split(",") if item.strip()]
    if not networks:
        raise ExampleError("NETLD_NETWORKS must contain at least one managed network.")
    return networks


@dataclass(frozen=True)
class Config:
    base_url: str
    api_key: str
    networks: list[str]
    output_path: Path
    failure_path: Path
    page_size: int
    scheme: str
    query: str

    @classmethod
    def from_env(cls, env_path: Path, output_base: Path) -> "Config":
        load_dotenv(env_path, override=True)
        base_url = os.getenv("NETLD_BASE_URL", "").strip()
        api_key = os.getenv("NETLD_API_KEY", "").strip()
        if not base_url:
            raise ExampleError("Set NETLD_BASE_URL in the environment file.")
        if not api_key:
            raise ExampleError("Set NETLD_API_KEY in the environment file.")
        try:
            page_size = int(os.getenv("NETLD_PAGE_SIZE", "500"))
        except ValueError as exc:
            raise ExampleError("NETLD_PAGE_SIZE must be a positive integer.") from exc
        if page_size <= 0:
            raise ExampleError("NETLD_PAGE_SIZE must be a positive integer.")

        def output_path(name: str, default: str) -> Path:
            value = Path(os.getenv(name, default))
            return value if value.is_absolute() else output_base / value

        return cls(
            base_url=base_url.rstrip("/"), api_key=api_key,
            networks=parse_networks(os.getenv("NETLD_NETWORKS", "Default")),
            output_path=output_path("NETLD_OUTPUT_FILE", "interfaces.csv"),
            failure_path=output_path("NETLD_FAILURE_FILE", "interface-failures.csv"),
            page_size=page_size,
            scheme=os.getenv("NETLD_SEARCH_SCHEME", "ipAddress").strip() or "ipAddress",
            query=os.getenv("NETLD_SEARCH_QUERY", ""),
        )


class NetLDClient:
    def __init__(self, base_url: str, api_key: str, timeout: float = 30):
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({"Authorization": f"Bearer {api_key}"})

    def _request(self, method: str, **kwargs: Any):
        try:
            response = self.session.request(
                method, f"{self.base_url}/rest", timeout=self.timeout,
                allow_redirects=False, **kwargs,
            )
        except requests.RequestException as exc:
            raise ExampleError(f"Could not reach {self.base_url}.") from exc
        if 300 <= response.status_code < 400:
            raise ExampleError(f"Request redirected to {response.headers.get('Location', '')}.")
        try:
            response.raise_for_status()
        except requests.RequestException as exc:
            raise ExampleError(f"Request failed with HTTP {response.status_code}.") from exc
        return response

    def login(self) -> None:
        self._request("GET")

    def call(self, method: str, **parameters: Any) -> Any:
        response = self._request(
            "POST", headers={"Content-Type": "application/json"},
            json={"jsonrpc": "2.0", "method": method, "params": parameters, "id": str(uuid.uuid4())},
        )
        try:
            data = response.json()
        except ValueError as exc:
            raise ExampleError(f"{method} returned invalid JSON.") from exc
        if data.get("error"):
            raise ExampleError(f"{method} failed: {json.dumps(data['error'])}")
        return data.get("result")

    def search_inventory(self, networks: list[str], scheme: str, query: str, offset: int, page_size: int):
        result = self.call(
            "Inventory.search", network=networks, scheme=scheme,
            query=query if query.endswith("\n") else f"{query}\n",
            pageData={"offset": offset, "pageSize": page_size},
            sortColumn="ipAddress", descending=False,
        )
        if not isinstance(result, dict):
            raise ExampleError("Inventory.search returned no page data.")
        return result

    def get_device_interfaces(self, network: str, ip_address: str):
        result = self.call("Inventory.getDeviceInterfaces", network=network, ipAddress=ip_address)
        if result is None:
            return []
        if not isinstance(result, list):
            raise ExampleError("Inventory.getDeviceInterfaces returned an invalid collection.")
        return result


def inventory_pages(client: Any, config: Config) -> Iterator[list[dict[str, Any]]]:
    offset = 0
    total: int | None = None
    while True:
        page = client.search_inventory(config.networks, config.scheme, config.query, offset, config.page_size)
        devices = page.get("devices") or []
        if not isinstance(devices, list):
            raise ExampleError("Inventory.search returned an invalid devices collection.")
        yield devices
        returned_page_size = int(page.get("pageSize") or config.page_size)
        if returned_page_size <= 0:
            raise ExampleError("Inventory.search returned an invalid page size.")
        if total is None and page.get("total") is not None:
            total = int(page["total"])
        if total is not None and offset + len(devices) >= total:
            return
        if total is None and len(devices) < returned_page_size:
            return
        if not devices:
            raise ExampleError("Inventory.search returned an empty page before the reported total.")
        offset += returned_page_size


def flatten_ip_addresses(interface: dict[str, Any]) -> str:
    values = []
    for address in interface.get("ipAddresses") or []:
        value = str(address.get("ipAddress") or "").strip()
        if not value:
            continue
        prefix = address.get("cidrPrefix")
        values.append(f"{value}/{prefix}" if prefix is not None else value)
    return ";".join(values)


def interface_row(device: dict[str, Any], interface: dict[str, Any]) -> dict[str, Any]:
    admin_up = interface.get("adminUp")
    if isinstance(admin_up, bool):
        admin_up = str(admin_up).lower()
    return {
        "network": device.get("network"), "deviceIpAddress": device.get("ipAddress"),
        "hostname": device.get("hostname"), "interfaceId": interface.get("id"),
        "interfaceIndex": interface.get("index"), "name": interface.get("name"),
        "ifName": interface.get("ifName"), "type": interface.get("type"),
        "description": interface.get("description"), "comment": interface.get("comment"),
        "macAddress": interface.get("macAddress"), "mtu": interface.get("mtu"),
        "speed": interface.get("speed"), "adminUp": admin_up,
        "vrfName": interface.get("vrfName"), "ipAddresses": flatten_ip_addresses(interface),
    }


def export_interfaces(client: Any, config: Config) -> tuple[int, int, int]:
    for destination in (config.output_path, config.failure_path):
        destination.parent.mkdir(parents=True, exist_ok=True)
    temporary_paths: list[Path] = []
    device_count = interface_count = failure_count = 0
    try:
        with tempfile.NamedTemporaryFile(
            mode="w", encoding="utf-8", newline="", dir=config.output_path.parent,
            prefix=f".{config.output_path.name}.", suffix=".tmp", delete=False,
        ) as interface_output, tempfile.NamedTemporaryFile(
            mode="w", encoding="utf-8", newline="", dir=config.failure_path.parent,
            prefix=f".{config.failure_path.name}.", suffix=".tmp", delete=False,
        ) as failure_output:
            temporary_paths = [Path(interface_output.name), Path(failure_output.name)]
            interface_writer = csv.DictWriter(interface_output, fieldnames=INTERFACE_FIELDS)
            failure_writer = csv.DictWriter(failure_output, fieldnames=FAILURE_FIELDS)
            interface_writer.writeheader()
            failure_writer.writeheader()
            for devices in inventory_pages(client, config):
                for device in devices:
                    device_count += 1
                    try:
                        interfaces = client.get_device_interfaces(device["network"], device["ipAddress"])
                    except Exception as exc:  # Preserve other devices when one lookup fails.
                        failure_writer.writerow({
                            "network": device.get("network"), "deviceIpAddress": device.get("ipAddress"),
                            "hostname": device.get("hostname"), "error": str(exc),
                        })
                        failure_count += 1
                        continue
                    for interface in interfaces:
                        interface_writer.writerow(interface_row(device, interface))
                        interface_count += 1
        os.replace(temporary_paths[0], config.output_path)
        os.replace(temporary_paths[1], config.failure_path)
        return device_count, interface_count, failure_count
    except Exception:
        for temporary_path in temporary_paths:
            temporary_path.unlink(missing_ok=True)
        raise


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--env", type=Path, default=Path(__file__).with_name(".env"))
    args = parser.parse_args()
    try:
        base = Path(__file__).resolve().parent
        config = Config.from_env(args.env, base)
        client = NetLDClient(config.base_url, config.api_key)
        client.login()
        devices, interfaces, failures = export_interfaces(client, config)
        print(f"Processed {devices} devices and wrote {interfaces} interfaces to {config.output_path}")
        print(f"Wrote {failures} device lookup failures to {config.failure_path}")
        return 2 if failures else 0
    except (ExampleError, OSError, ValueError, KeyError) as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
